add keyctl support for loading LUKS passphrase into a keyring

maciejsszmigiero opened the pull request:
cryptsetup LUKS2 format comes with an ability to automatically unlock
multiple devices (root, swap, etc.) sharing the same passphrase, without
retyping it for each of them, by loading it into the user keyring.

This commit adds such (optional) genkernel support for loading LUKS
passphrase into the user keyring on boot.

In the default mode of operation the newly added key is (possibly) used
only to unlock root and swap devices and is removed soon after that.
By providing appropriate kernel command line parameter the key can be left
in the keyring instead (with an optional timeout) for unlocking other LUKS
devices post-initramfs time.

Because one of the most common use cases of this functionality will be
having an encrypted swap for doing suspend to disk (hibernation) let's also
make sure that we don't unlock the root device when doing so is unnecessary
(when we are resuming the system from hibernation).

Since the security of a FDE passphrase is of paramount importance in this
solution significant care has been taken not to leak it accidentally:
* The passphrase is read directly by keyctl to avoid storing it in the
shell,

* If the passphrase is used only to unlock root and swap devices (which is
the default mode of operation) the init script will check whether its
removal from keyring has actually succeeded and, if not, reboot the system
rather than continue while leaving it exposed,

* keyutils includes a patch (already upstreamed) to wipe the passphrase
from memory when no longer needed.

maciejsszmigiero commented:
Rebased onto the current master.

alxkv commented:
Hi!
Is this patch supposed to work together with systemd's crypttab?
If yes then it is exactly what I was looking for and want to test it

maciejsszmigiero committed:
genkernel: add keyctl support for loading LUKS passphrase into a keyring

cryptsetup LUKS2 format comes with an ability to automatically unlock
multiple devices (root, swap, etc.) sharing the same passphrase, without
retyping it for each of them, by loading it into the user keyring.

This commit adds such (optional) genkernel support for loading LUKS
passphrase into the user keyring on boot.

In the default mode of operation the newly added key is (possibly) used
only to unlock root and swap devices and is removed soon after that.
By providing appropriate kernel command line parameter the key can be left
in the keyring instead (with an optional timeout) for unlocking other LUKS
devices post-initramfs time.

Because one of the most common use cases of this functionality will be
having an encrypted swap for doing suspend to disk (hibernation) let's also
make sure that we don't unlock the root device when doing so is unnecessary
(when we are resuming the system from hibernation).

Since the security of a FDE passphrase is of paramount importance in this
solution significant care has been taken not to leak it accidentally:
* The passphrase is read directly by keyctl to avoid storing it in the
shell,

* If the passphrase is used only to unlock root and swap devices (which is
the default mode of operation) the init script will check whether its
removal from keyring has actually succeeded and, if not, reboot the system
rather than continue while leaving it exposed,

* keyutils includes a patch (already upstreamed) to wipe the passphrase
from memory when no longer needed.

Signed-off-by: Maciej S. Szmigiero <mail@maciej.szmigiero.name>

maciejsszmigiero commented:
Rebased the patch onto the current master.

> Is this patch supposed to work together with systemd's crypttab?
> If yes then it is exactly what I was looking for and want to test it

I'm actually using it with OpenRC but I don't see any reason why this wouldn't work with systemd as long as it doesn't force a password prompt for LUKS volumes that can be unlocked using a keyring token.

Actually, I see a call to `crypt_activate_by_token()` in systemd here: https://github.com/systemd/systemd/blob/07ccf434e77b17b20d773ce8b2637083cd4fdafc/src/cryptsetup/cryptsetup.c#L945
So it looks like unlocking later LUKS volumes using genkernel-provided passphrase loaded into the user keyring should work (haven't tested it myself, though).

This will need the `keyctl_keykeep` kernel command line option to make sure the passphrase is not removed after unlocking root and swap devices in genkernel (`keyctl_keytimeout=<timeout>` can be specified, too, to remove it after few minutes).

maciejsszmigiero commented:
Rebased this PR onto the current master.

@Whissi I see that you are doing some genkernel work recently.
Can we have this reviewed? Thanks.

I have been using this patch to avoid typing LUKS passphrase multiple times on boot for 1.5 years now.

Whissi commented:
Hi,

wasn't there a bug where I already commented on this? I have the following problem with this addition:

- It's against best practice. You should never re-use same passphrase for multiple volumes.
- Current PR will create keyslot and kill it before transferring control from initramfs to real system so I really don't get the value of this at all.
- The correct way of handling multiple volumes in genkernel is having one prompt for root volume where you store keyfiles for additional volumes which will be used by real system to unlock these additional volumes after taking control.

maciejsszmigiero commented:
Thanks for your comments.

> wasn't there a bug where I already commented on this? 

Don't remember anything like that.

>  * It's against best practice. You should never re-use same passphrase for multiple volumes.

Yes, there is a certain usability / security tradeoff there.
People have memory only limited number of passphrases.

>  * Current PR will create keyslot and kill it before transferring control from initramfs to real system
> so I really don't get the value of this at all.

By default the code will remove the newly added keyring key (passphrase) before starting the real init.
This is designed so root and swap volumes can share the same passphrase.

If one adds keyctl_keykeep and (for example) keyctl_keytimeout=60 other volumes can be unlocked by init service scripts later during boot, and the passphrase will be removed from memory after that.

> * The correct way of handling multiple volumes in genkernel is having one prompt for root volume where you store
> keyfiles for additional volumes which will be used by real system to unlock these additional volumes after taking control.

This does not work for resuming from suspend to disk use case where we have to unlock only swap and cannot safely mount root fs.

maciejsszmigiero commented:
Rebased the patch onto the current master.

It's worth noting that in addition to the above using an additional volume with keyfiles isn't compatible with native Linux tools (`cryptsetup`, `UDisks`, etc.), which do not know how to open such volume on their own.

Since this patch solves real use cases (resuming from hibernation where the root fs cannot be safely mounted and automatic unlocking other volumes by init scripts while removing passphrase from memory soon after that) and it has been waiting for 2+ years already now, can we have this finally merged @Whissi ?
Thanks.

duxsco commented:
@Whissi With [decrypt_keyctl](https://sources.debian.org/src/cryptsetup/2%3A2.4.3-1/debian/scripts/decrypt_keyctl/), Debian uses a similar approach as in this PR for caching passphrases and unlocking multiple LUKS volumes, however, without the support for root LUKS volumes. In Arch Linux, [rd.luks parameters can be specified multiple times to unlock multiple LUKS encrypted volumes](https://wiki.archlinux.org/title/Sd-encrypt#Using_sd-encrypt_hook) including root LUKS volumes.

Unlocking multiple root LUKS volumes with a single passphrase prompt is a feature I dearly miss on Gentoo. This would come in handy in setups with multiple disks with LUKS on top of the partitions and filesystem-level RAID on top of LUKS (BTRFS/ZFS), as shown with the 4th partition:

```
PC∕Laptop──────────────────────────┐
└── ∕dev∕sda                       └── ∕dev∕sdb
    ├── 1. EFI System Partition        ├── 1. EFI System Partition
    ├── 2. MDADM RAID 1                ├── 2. MDADM RAID 1
    │   └── LUKS                       │   └── LUKS
    │       └── Btrfs                  │       └── Btrfs
    │           └── boot               │           └── boot
    ├── 3. LUKS                        ├── 3. LUKS
    │   └── MDADM RAID 1               │   └── MDADM RAID 1
    │       └── SWAP                   │       └── SWAP
    └── 4. LUKS ("system" partition)   └── 4. LUKS ("system" partition)
        └── BTRFS (raid1)                  └── BTRFS (raid1)
            └── subvolume                      └── subvolume
                ├── @binpkgs                       ├── @binpkgs
                ├── @distfiles                     ├── @distfiles
                ├── @home                          ├── @home
                ├── @ebuilds                       ├── @ebuilds
                └── @root                          └── @root
```

There is [this bug report](https://bugs.gentoo.org/694778) which helped me to write [my own set of genkernel patches](https://github.com/duxsco/gentoo-genkernel-patches/). It would be great for Gentoo to support multi-volume root LUKS unlock and let the user decide which approach to take: `initramfs+keyfile` (Arch Linux's approach) or `keyctl` (Debian's approach). As we know, freedom of choice is Gentoo's way 😉

maciejsszmigiero commented:
Thanks David for your input.

It's worth noting that with this PR other volumes than root and swap can be automatically unlocked, too, by init service scripts later during boot by adding a `keyctl_keykeep` parameter with (for example) `keyctl_keytimeout=60` parameter.
The passphrase will be removed by kernel from its memory soon after that.

duxsco commented:
@maciejsszmigiero Thanks for taking the initiative to push this forward.

Another way to "ease" the LUKS unlock procedure could be the use of a [GnuPG smartcard and reader incl. pinpad](https://github.com/dracutdevs/dracut/blob/master/modules.d/91crypt-gpg/README) which is something I consider realising on my laptop with my [GnuPG smartcard](https://github.com/duxsco/gpg-smartcard). You have to type in the pin multiple times though. But, this shouldn't bother much for short pins with the safety of the smartcard becoming unusable after 3 failed entries.

robbat2 commented:
@Whissi i'm inclined to merge this unless you have strong objections, because it solves use cases that aren't otherwise possible

maciejsszmigiero commented:
Will provide a rebased version tomorrow.

maciejsszmigiero commented:
Rebased the patch onto the current master and updated the used `keyutils` version to 1.6.3 (the current latest).

This PR good for merging now @robbat2 before some other patch creates a new merge conflict.

robbat2 reviewed (commented):
approved

robbat2 commented:
@maciejsszmigiero  thanks for this.
Can you also please check https://github.com/gentoo/genkernel/pull/36 and see if it needs updates for interactions?

maciejsszmigiero commented:
Thanks for merging this PR @robbat2 .

Will review the related parts of #36 this week.

FlyingWaffleDev commented on defaults/linuxrc:
```diff
@@ -656,10 +665,23 @@ then
 if [ "${CDROOT}" != '1' ]
 then
-	start_LUKS
+	if ( [ -n "${CRYPT_SWAP_KEY}" ] && [ -z "${CRYPT_SWAP_KEYDEV}" ] ) || \
+	   ( [ -n "${CRYPT_SWAP_HEADER}" ] && [ -z "${CRYPT_SWAP_HEADERDEV}" ] )
+	then
+		# the swap key or header might be on the root fs so start it first in this case
+		start_LUKS_root
+		luks_root_started=1
+		start_LUKS_swap
+	else
+		# we don't need to start the root at all if we are resuming from suspend
+		start_LUKS_swap
+	fi
+
```
Some assumptions are made about configuration in this code block.  You aren't accounting for: swapfile on root, root and swap in lvm volume, etc.  These would all need LUKS_root opened before running do_resume.  Technically I suppose you could define a crypt_swap_key value the same as your root value, but that seems like misleading configuration.

Could maybe add:
`( [ "${REAL_ROOT}" = "${REAL_RESUME}" ]  || [ ${USE_LVM_NORMAL} -eq 1 ] )`

This would solve both cases (I think).

maciejsszmigiero replied:
> swapfile on root,

Looking at the relevant code in `swsusp_resume()` from `defaults/initrd.scripts` only swap partitions are supported for resume from hibernation.

You might be correct with respect to LVM though.

FlyingWaffleDev replied:
I beg to differ, I'm currently running a 'LUKS -> btrfs -> swapfile' setup which resumes from hibernation just fine.  I just patched out the code-block above to enforce root LUKS being opened before resume is run, and I have 'resume=' and 'resume_offset=' set in my kernel cmdline.

maciejsszmigiero replied:
I see, you are using genkernel to write the resume device path to `/sys/power/resume` but at the same time providing the kernel directly with the swap file offset via the `resume_offset` kernel parameter. 

Will you open a PR with your proposed change then?

FlyingWaffleDev replied:
Sure, I can do that.
